### ISSUE TITLE
Pre-trained TalkNet checkpoint only achieves ~20 % mAP on UniTalk val split

### ISSUE THREAD

**Issue by Aubeisme** [plnguyen2908/UniTalk-ASD-code#3]
I directly used the repository [TalkNet-ASD](https://github.com/TaoRuijie/TalkNet-ASD) to load the pretrained model(https://drive.google.com/file/d/1Uq8vp__7UywtdY0z5zz8QtcbEwFfXlNw/view?usp=sharing)(you shared)and perform evaluation. My steps were:
Downloaded the UniTalk dataset via
python download_dataset.py --save_path /path/to/the/dataset.
Converted UniTalk’s val_orig.csv into AVA-style loader files (e.g., loader.csv) so that TalkNet-ASD’s AVA evaluation script could read them.
Ran the TalkNet evaluation pipeline.
However, the final mAP is far below the expected value.
Could you tell me which of these steps is likely to be wrong or what I might have missed? Thanks

**Comment by plnguyen2908** [comment]
Hi, we used the original TalkNet codebase. Here is the code that we use to convert val_orig.csv to val_loader.csv:

```
import pandas as pd
import random
import os

fps = 25.0
input_csv  = "/path/to/csv/val_orig.csv"      
output_csv = "/path/to/csv/val_loader.csv"   # should be the same path as val_orig.csv

df = pd.read_csv(input_csv)

random.seed(42) 
with open(output_csv, "w") as fout:
    for entity_id, grp in df.groupby("entity_id"):
        count    = len(grp)
        labels   = grp["label_id"].tolist()
        rnd_id   = random.randint(1, 1_000_000)
        fout.write(f"{entity_id}\t{count}\t{fps}\t{labels}\t{rnd_id}\n")

```
Can you try it and let me know if things work?

**Comment by Aubeisme** [comment]
Thanks, I'll try it right now.
________________________________
发件人: Le Thien Phuc Nguyen ***@***.***>
发送时间: 2025年7月30日 14:09
收件人: plnguyen2908/UniTalk-ASD-code ***@***.***>
抄送: RuiJin SHA ***@***.***>; Author ***@***.***>
主题: Re: [plnguyen2908/UniTalk-ASD-code] Pre-trained TalkNet checkpoint only achieves ~20 % mAP on UniTalk val split (Issue #3)

[https://avatars.githubusercontent.com/u/86179608?s=20&v=4]plnguyen2908 left a comment (plnguyen2908/UniTalk-ASD-code#3)<https://github.com/plnguyen2908/UniTalk-ASD-code/issues/3#issuecomment-3136529169>

Hi, we used the original TalkNet codebase. Here is the code that we use to convert val_orig.csv to val_loader.csv:

import pandas as pd
import random
import os

fps = 25.0
input_csv = "/path/to/csv/val_orig.csv"
output_csv = "/path/to/csv/val_loader.csv" # should be the same path as val_orig.csv

df = pd.read_csv(input_csv)

random.seed(42)
with open(output_csv, "w") as fout:
for entity_id, grp in df.groupby("entity_id"):
count = len(grp)
labels = grp["label_id"].tolist()
rnd_id = random.randint(1, 1_000_000)
fout.write(f"{entity_id}\t{count}\t{fps}\t{labels}\t{rnd_id}\n")

Can you try it and let me know if things work?

―

**Comment by plnguyen2908** [comment]
for your information, i have rerun the evaluation and got:

<img width="686" height="106" alt="Image" src="https://github.com/user-attachments/assets/2c4cce22-7654-47ee-aa50-d6a9124be8d3" />

Hopefully this will help

**Comment by plnguyen2908** [comment]
feel free to reopen when you need help

**Comment by Aubeisme** [comment]
After using the code you provided, I reran and got:
[cid:165efa82-60ae-40fa-b8f7-2c089a51b406]
I guess there's an issue with my dataset, so I think I'll re-download it. I'll get back to you later. Thanks
________________________________
发件人: Le Thien Phuc Nguyen ***@***.***>
发送时间: 2025年7月30日 14:23
收件人: plnguyen2908/UniTalk-ASD-code ***@***.***>
抄送: RuiJin SHA ***@***.***>; Author ***@***.***>
主题: Re: [plnguyen2908/UniTalk-ASD-code] Pre-trained TalkNet checkpoint only achieves ~20 % mAP on UniTalk val split (Issue #3)

[https://avatars.githubusercontent.com/u/86179608?s=20&v=4]plnguyen2908 left a comment (plnguyen2908/UniTalk-ASD-code#3)<https://github.com/plnguyen2908/UniTalk-ASD-code/issues/3#issuecomment-3136592810>

feel free to reopen when you need help

―

**Comment by Aubeisme** [comment]
Hi,I re-downloaded the UniTalk dataset, recreated the TalkNet environment, used the code you provided earlier to generate loader.csv, and also used the pre-trained model you provided. However, the result is still only a 20% mAP. I don't know exactly where I went wrong, and I really hope to get your help. Thanks.

<img width="1851" height="109" alt="Image" src="https://github.com/user-attachments/assets/d47194a1-514c-41eb-ab60-086a81581e8c" />

**Comment by plnguyen2908** [comment]
Hi, 

I am investigating the problem. Will reach back to you soon.

**Comment by plnguyen2908** [comment]
Hi,

I have debugged the problem, here is the code to generate the loader:

```
import pandas as pd
import random
import os

fps = 25.0
input_csv  = "/nobackup/le/UniTalk_asd_download/csv/val_orig.csv"      
output_csv = "/nobackup/le/UniTalk_asd_download/csv/val_loader.csv"   

df = pd.read_csv(input_csv)

random.seed(42) 
with open(output_csv, "w") as fout:
    for entity_id, grp in df.groupby("entity_id", sort=False):
        count    = len(grp)
        labels   = grp["label_id"].tolist()
        rnd_id   = random.randint(1, 1_000_000)
        # Viết ra file, dùng '\t' làm phân tách
        fout.write(f"{entity_id}\t{count}\t{fps}\t{labels}\t{rnd_id}\n")

```

Basically, you just need to add sort=False in df.groupby

**Comment by Aubeisme** [comment]
Thank you. It works. Wish you every success in your academic research.

<img width="1091" height="57" alt="Image" src="https://github.com/user-attachments/assets/38f50511-f0b4-4481-a5a7-1f872ea9edec" />

**Comment by plnguyen2908** [comment]
Thank you for your kind words! Wish you the best, too.
